source/material_model/rheology/elasticity.cc

```diff
@@ -252,11 +252,33 @@ namespace aspect
             for (unsigned int i=0; i < in.n_evaluation_points(); ++i)
               {
                 // Get old stresses from compositional fields
+                // TODO: use $tau^{t+\Delta te}$ or evaluate the old solution for stress
+                // (and viscosity)
+                // Explanation: for the assembly of the Stokes equations (the force term goes
+                // into the rhs of the momentum equation), this function is evaluated.
+                // This is after the advection equations have been solved, and hence in.composition
+                // contains the rotated, averaged and advected stresses at $tau^{t+\Delta t}$.
+                //
+                // For an iterative Advection scheme, in.composition could even be different
+                // every time the Stokes system is assembled.
+                //
+                // Moresi et al. (2003) use the full stresses at $t$ (Eq.  30) in the force term.
+                // There is also an incorrect minus sign in the force term.
```

Comment:
Is this incorrect minus sign in the Moresi paper?

Reply:
@naliboff in the Moresi paper, the force term is on the LHS (equations 2, 31). In ASPECT, the force term is on the RHS. We both use the same sign (i.e. - eta_eff/eta_el * [some tau]). So if we are interpreting the stress terms correctly, there must be a sign error.

Reply:
We also need to use $tau^{t+\Delta te}$ to calculate the effective viscosity on the yield surface *for a non-damped elastic element* (l244 of rheology/visco_plastic.cc, eq. 36 of Moresi et al. 2003). This is also obtained from in.composition[i,j] (l114 of rheology/visco_plastic.cc).

Reply:
On a related note, I think there should only be one elastic element at any evaluation point, rather than one-per-compositional-field-per-element. In other words, we should drag the elasticity calculations outside the loop over compositional fields and use a volume-averaged shear modulus. There isn't an obvious choice for whether we use harmonic (Reuss, isostress) or arithmetic (Voigt, isostrain) averaging for elasticity. I have a slight preference for arithmetic (probably better for small-strains).

Reply:
> We also need to use $tau^{t+\Delta te}$ (rotated, advected) to calculate the effective viscosity on the yield surface _for a non-damped elastic element_ (l244 of rheology/visco_plastic.cc, eq. 36 of Moresi et al. 2003). This is also obtained from in.composition[i,j] (l114 of rheology/visco_plastic.cc).

Just to be sure, the stress in in.composition is only $tau^{t+\Delta te}$ when $\Delta te$ = $\Delta t$, right? Otherwise it would be $tau^{t+\Delta t}$.

Reply:
When we evaluate the material model to compute the rhs force term, `in.composition` has been updated using the reaction_outputs. 

At the moment, Advection iterations are not allowed because this could indeed lead to a different `in.composition` in each iteration.

Reply:
The force term could still be different during every non-linear iteration, because the velocity that goes into the Advection equation is updated. Already for single Advection schemes, the viscosity will be different in every Stokes iteration.